tool/tbot/main.go

```diff
@@ -0,0 +1,631 @@
+		// TODO: error if duration < renewalInterval? next renewal attempt will fail
+
+		// Immediately attempt to reconnect using the new identity (still
+		// haven't persisted the known-good certs).
+		newClient, err := authenticatedUserClientFromIdentity(newIdentity, cfg.AuthServer)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Attempt a request to make sure our client works.
+		// TODO: consider a retry/backoff loop.
+		if _, err := newClient.Ping(context.Background()); err != nil {
+			return trace.WrapWithMessage(err, "unable to communicate with auth server")
+		}
+
+		log.Debug("Auth client now using renewed credentials.")
+		client = newClient
+		ident = newIdentity
+
+		// Now that we're sure the new creds work, persist them.
+		if err := identity.SaveIdentity(newIdentity, botDestination); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Determine the default role list based on the bot role. The role's
+		// name should match the certificate's Key ID (user and role names
+		// should all match bot-$name)
+		defaultRoles, err := fetchDefaultRoles(context.Background(), client, ident.Cert.KeyId)
+		if err != nil {
+			log.WithError(err).Warnf("Unable to determine default roles, no roles will be requested if unspecified")
+			defaultRoles = []string{}
+		}
+
+		// Next, generate impersonated certs
+		expires := time.Unix(int64(ident.Cert.ValidBefore), 0)
+		for _, dest := range cfg.Destinations {
+			destImpl, err := dest.GetDestination()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			var desiredRoles []string
+			if len(dest.Roles) > 0 {
+				desiredRoles = dest.Roles
+			} else {
+				log.Debugf("Destination specified no roles, defaults will be requested: %v", defaultRoles)
+				desiredRoles = defaultRoles
+			}
+
+			impersonatedIdent, err := generateImpersonatedIdentity(client, ident, expires, desiredRoles)
+			if err != nil {
+				return trace.WrapWithMessage(err, "Failed to generate impersonated certs for %s: %+v", destImpl, err)
+			}
+
+			impersonatedIdentStr, err := describeIdentity(impersonatedIdent)
+			if err != nil {
+				return trace.WrapWithMessage(err, "could not describe impersonated certs for destination %s", destImpl)
+			}
+			log.Infof("Successfully renewed impersonated certificates for %s, %s", destImpl, impersonatedIdentStr)
+
+			if err := identity.SaveIdentity(impersonatedIdent, destImpl); err != nil {
+				return trace.WrapWithMessage(err, "failed to save impersonated identity to destination %s", destImpl)
+			}
+
+			for _, templateConfig := range dest.Configs {
+				template, err := templateConfig.GetConfigTemplate()
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				if err := template.Render(client, impersonatedIdent, dest); err != nil {
+					log.WithError(err).Warnf("Failed to render config template %+v", templateConfig)
+				}
+			}
+		}
+
+		log.Infof("Persisted new certificates to disk. Next renewal in approximately %s", cfg.RenewInterval)
+		<-ticker.C
```

Comment:
Should this select on a context to be able to cancel?

Reply:
Yes, that'll be useful for graceful shutdown support as well.